inst/tutorials/practical0-1/practical0-1.Rmd

```diff
@@ -0,0 +1,444 @@
+in a later practical, but for now we'll focus on the simplest - something that
+tells you to do something a fixed number of times that you know in advance.
+This is called a `for` loop.
+
+A `for` loop always has the same style:
+
+```{r, eval=FALSE}
+for (one in many) {
+  # Do something
+    ... exciting code maybe using the variable
+        one (but it doesn't have to) ...
+}
+```
+
+It looks a bit like a function, but what it does is set the variable one to
+each of the values in many one at a time, and then does something. Here are two
+very simple examples to get an idea of what it might do:
+
+```{r simple_loop, exercise.cap = "Example 4", exercise=TRUE}
+words <- c("Hello", "Goodbye")
+for (word in words) {
+  print(word)
+}
+```
+
+Try changing the `words` variable. You'll see that the loop runs once for each
+word in the `words` vector. We can use this to do something a fixed number of
+times without necessarily caring about the entries:
+
+```{r times_loop, exercise.cap = "Example 5", exercise=TRUE}
+# Make a sequence of 7 numbers from 1 to 7
+times <- seq(from = 1, to = 7) # This is the same as 1:7 or seq(1, 7)
+
+# Print "I'm in a loop" repeatedly
+for (iteration in times) {
+  print("I'm in a loop")
+}
+```
+
+Change the length of the `times` vector. You'll see that it prints out the
+text once for each element in `times` even though it never looks to see what
+`iteration` is (it doesn't care).
+
+Now look at this code:
+
+```{r no_loop, exercise.cap = "Exercise 4", exercise=TRUE}
+# A function to add one to a number
+add_one <- function(number)
+  number + 1
+
+# Initialise the value
+current.value <- 0
+
+current.value <- add_one(current.value)
+current.value <- add_one(current.value)
+current.value <- add_one(current.value)
+current.value <- add_one(current.value)
+current.value <- add_one(current.value)
+
+# Output the updated current.value at the end to see what happened
+current.value
+```
+
+```{r no_loop-check}
+grade_result(pass_if(~identical(.result, 10)))
+```
+
+It just runs our `add_one()` function from earlier five times to update the
+variable `current.value` to 5. Change the code so it does this ten times to
+produce 10.
+
+Now look at this next piece of code:
+
+```{r with_loop, exercise.cap = "Exercise 5", exercise=TRUE}
+# A function to add one to a number
+add_one <- function(number) {
+  number + 1
+}
+
+# Initialise the value
+current.value <- 0
+
+# This now is a loop!
+for (loop in 1:10) {
+  current.value <- add_one(current.value)
+}
+
+# Output the updated current.value at the end to see what happened
+current.value
+```
+
+```{r with_loop-check}
+grade_result(pass_if(~identical(.result, 20)))
+```
+
+It does the same thing - it calls `add_one()` ten times, but it's much shorter,
+which generally makes it easier to see what's going on. Now change the code
+to call `add_one()` 100 times. Hopefully you can see that this is really
+starting to help now!
+
```

Comment:
The next step might be to `print(paste("Loop", loop))` at line 417 with a description here, emphasising the points made in the earlier examples

Reply:
I've added this afterwards.